adapters/pubmatic/pubmatic_test.go

```diff
@@ -359,3 +395,120 @@ func TestPubmaticAdapter_MakeBids(t *testing.T) {
+			args: args{
+				bidRequest: &openrtb2.BidRequest{Ext: json.RawMessage(`{"prebid":{"alternatebiddercodes":{"enabled":false,"bidders":{"pubmatic":{"enabled":true,"allowedbiddercodes":["groupm"]}}}}}`)},
+			},
+			want: []string{"pubmatic"},
+		},
+		{
+			name: "alternatebiddercodes disabled at bidder level",
+			args: args{
+				bidRequest: &openrtb2.BidRequest{Ext: json.RawMessage(`{"prebid":{"alternatebiddercodes":{"enabled":true,"bidders":{"pubmatic":{"enabled":false,"allowedbiddercodes":["groupm"]}}}}}`)},
+			},
+			want: []string{"pubmatic"},
+		},
+		{
+			name: "alternatebiddercodes list not defined",
+			args: args{
+				bidRequest: &openrtb2.BidRequest{Ext: json.RawMessage(`{"prebid":{"alternatebiddercodes":{"enabled":true,"bidders":{"pubmatic":{"enabled":true}}}}}`)},
+			},
+			want: []string{"all"},
+		},
+		{
+			name: "wildcard in alternatebiddercodes list",
+			args: args{
+				bidRequest: &openrtb2.BidRequest{Ext: json.RawMessage(`{"prebid":{"alternatebiddercodes":{"enabled":true,"bidders":{"pubmatic":{"enabled":true,"allowedbiddercodes":["*"]}}}}}`)},
+			},
+			want: []string{"all"},
+		},
+		{
+			name: "empty alternatebiddercodes list",
+			args: args{
+				bidRequest: &openrtb2.BidRequest{Ext: json.RawMessage(`{"prebid":{"alternatebiddercodes":{"enabled":true,"bidders":{"pubmatic":{"enabled":true,"allowedbiddercodes":[]}}}}}`)},
+			},
+			want: []string{"pubmatic"},
+		},
+		{
+			name: "only groupm in alternatebiddercodes allowed",
+			args: args{
+				bidRequest: &openrtb2.BidRequest{Ext: json.RawMessage(`{"prebid":{"alternatebiddercodes":{"enabled":true,"bidders":{"pubmatic":{"enabled":true,"allowedbiddercodes":["groupm"]}}}}}`)},
+			},
+			want: []string{"pubmatic", "groupm"},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var reqExt *openrtb_ext.ExtRequest
+			if len(tt.args.bidRequest.Ext) > 0 {
+				err := json.Unmarshal(tt.args.bidRequest.Ext, &reqExt)
+				if err != nil {
+					t.Errorf("getAlternateBidderCodesFromRequest() = %v", err)
+				}
+			}
+
+			got := getAlternateBidderCodesFromRequestExt(reqExt)
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("getAlternateBidderCodesFromRequest() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func Test_assertBidderCodes(t *testing.T) {
+	type args struct {
+		bidders []string
+	}
+	tests := []struct {
+		name string
+		args args
+		want []string
+	}{
+		{
+			name: "nil input",
+			args: args{
+				bidders: nil,
+			},
+			want: nil,
+		},
+		{
+			name: "input with duplicate, uppercase and whitespace entries",
+			args: args{
+				bidders: []string{"Pubmatic", "Groupm", "groupm", " ", "", " appnexus "},
+			},
+			want: []string{"groupm", "appnexus"},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := assertBidderCodes(tt.args.bidders); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("assertBidderCodes() = %v, want %v", got, tt.want)
+			}
```

Comment:
I think it would be better to use the testify package instead of reflect. You could do the following:
```
result := assertBidderCodes(tt.args.bidders)
assert.ElementsMatch(t, result, tt.want, tt.name)
```

Reply:
removed the comment w.r.t above comments